### PR TITLE
freeCompilerArgs in tasks.KotlinCompile

### PR DESCRIPTION
**Pull request by Akshay0701**
## Purpose / Description

A list of additional compiler arguments

## Fixes
[comment](https://github.com/ankidroid/Anki-Android/pull/10060#discussion_r792913468)

_Links to blog posts, patterns, libraries or addons used to solve this problem_

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by david-allison-1 — COMMENTED** [review]
Ideally the commit message should explain the flag and why the `all` value was chosen. Maybe some of this explanation could be in a comment with a link to the docs

**Review by mikehardy — CHANGES_REQUESTED** [review]
@Akshay0701 this looks like it could go in but I agree with this in principle:

> Ideally the commit message should explain the flag and why the all value was chosen. Maybe some of this explanation could be in a comment with a link to the docs

To be most specific, as I'm maintaining gradle files here a lot, I want it in a comment with a link to the docs. I am certain I'll click that in the future when making some change...

**Comment by Akshay0701** [comment]
sorry, for the delay

**Review by david-allison — CHANGES_REQUESTED** [review]
> I want it in a comment with a link to the docs. I am certain I'll click that in the future when making some change...

Could this be in a code comment as well as the commit message? Message looks good

**Comment by Akshay0701** [comment]
okay

**Review by david-allison — APPROVED** [review]
LGTM

File: AnkiDroid/build.gradle [hunk 1]
```diff
@@ -238,8 +238,22 @@ task installGitHook(type: Copy) {
...
 tasks.getByPath(':AnkiDroid:preBuild').dependsOn installGitHook
 
+/**
+    The Kotlin compilation task names in Android projects contain the build variant names (those are combined from build type, product flavor and other settings and look like debug or releaseUnitTest -- the tasks are compileDebugKotlin and compileReleaseUnitTestKotlin respectively).
+    In kotlinOptions freeCompilerArgs(list of additional compiler arguments) is a attribute,
+    https://kotlinlang.org/docs/gradle.html#attributes-common-to-jvm-js-and-js-dce
+
+    There's no freeCompilerArgs in our kotlinCompile, we want to configure and to do that, we used freeCompilerArgs = ['-Xjvm-default=all'],
+    https://kotlinlang.org/docs/gradle.html#compiler-options
+
```
Comment (mikehardy) [inline AnkiDroid/build.gradle:248]: I'm sorry, but this comment is not illuminating at all - I had no idea what was going on here when I went to review it.
So I chased down the docs.

None of the part about task naming is important to understanding the build.gradle setting
Of course there was no freeCompilerArgs before, that's why we're adding it, but future reviewers won't care whether it was here before or not, they will want to know *what it does*

What it does is enable Kotlin<->Java interop for concrete implementations of functions inside Kotlin interfaces.

So the important link and comment is explaining that and linking to the interop documentation

```suggestion
Kotlin allows concrete function implementations inside interfaces.
For those to work when Kotlin compilation targets the JVM backend, you have to enable the interoperability via
'freeCompilerArgs' in your gradle file, and you have to choose one of the appropriate '-Xjvm-default' modes.

https://kotlinlang.org/docs/java-to-kotlin-interop.html#default-methods-in-interfaces

```

**Review by mikehardy — CHANGES_REQUESTED** [review]
I was totally confused by the comment personally, I couldn't figure out *why*.
What do you think of my suggestion?

File: AnkiDroid/build.gradle (hunk 1 above)
Comment (Akshay0701) [reply AnkiDroid/build.gradle:248]: Thanks Mike for improving comments

**Review by mikehardy — APPROVED** [review]
documentation is harder than code, in many ways 😅 - LGTM!

**Pull request merged**